mautic/mautic
# WIP: Improve performance of various queries used in reports

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [X]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

On a site we manage, we are seeing continuous 504s due to the dashboard and reporting pages' queries taking too long. 30+ seconds isn't a rarity. We noticed MySQL was using either the wrong index, or no index at all. These changes address that issue.

These are the changes we've undertaken so far to improve performance.

- Remove the "code" index from the page_hits table. MySQL loves that index, and as long as it was present MySQL refused to use the correct index.
- Added 1 index to campaign_lead_event_log table, and code changes in CampaignBundle/LeadEventLogRepository to force using the correct index
- Added 3 indices to lead_event_log table, and code changes in LeadBundle/LeadEventLogRepository  to force using the correct ones
- Updated TimelineTrait, to add an extra option to enforce a given index
- Updated ReportSubscriber, to ensure leftJoins are only added when actually necessary
- Created generated column, and use that in ChartQuery instead of using TIMESTAMPDIFF in the WHERE clause, which prevents MySQL from using an index



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Check dashboard and premade Reporting pages. Should have no errors.

But to properly test it, we would need a large enough dataset.
